Tools/wasm/wasm_build.py

```diff
@@ -0,0 +1,549 @@
+https://emscripten.org/docs/getting_started/downloads.html how to install
+Emscripten and how to activate the SDK with ". /path/to/emsdk/emsdk_env.sh".
+
+    git clone https://github.com/emscripten-core/emsdk.git /path/to/emsdk
+    cd /path/to/emsdk
+    ./emsdk install latest
+    ./emsdk activate latest
+    source /path/to/emsdk_env.sh
+"""
+
+INSTALL_WASI_SDK = """
+wasm32-wasi builds need WASI SDK. Please fetch the latest SDK from
+https://github.com/WebAssembly/wasi-sdk/releases and install it to
+"/opt/wasi-sdk". Alternatively you can install the SDK in a different location
+and point the environment variable WASI_SDK_PATH to the root directory
+of the SDK. The SDK is available for Linux x86_64, macOS x86_64, and MinGW.
+"""
+
+INSTALL_WASMTIME = """
+wasm32-wasi tests require wasmtime on PATH. Please follow instructions at
+https://wasmtime.dev/ to install wasmtime.
+"""
+
+
+def get_emscripten_root(emconfig: pathlib.Path = EM_CONFIG) -> pathlib.Path:
+    """Parse EM_CONFIG file and lookup EMSCRIPTEN_ROOT
+
+    The ".emscripten" config file is a Python snippet that uses "EM_CONFIG"
+    environment variable. EMSCRIPTEN_ROOT is the "upstream/emscripten"
+    subdirectory with tools like "emconfigure".
+    """
+    if not emconfig.exists():
+        return _MISSING
+    with open(emconfig, encoding="utf-8") as f:
+        code = f.read()
+    # EM_CONFIG file is a Python snippet
+    local = {}
+    exec(code, globals(), local)
+    return pathlib.Path(local["EMSCRIPTEN_ROOT"])
+
+
+EMSCRIPTEN_ROOT = get_emscripten_root()
+
+
+class ConditionError(ValueError):
+    def __init__(self, info: str, text: str):
+        self.info = info
+        self.text = text
+
+    def __str__(self):
+        return f"{type(self).__name__}: '{self.info}'\n{self.text}"
+
+
+class MissingDependency(ConditionError):
+    pass
+
+
+class DirtySourceDirectory(ConditionError):
+    pass
+
+
+@dataclasses.dataclass
+class Platform:
+    """Platform-specific settings
+
+    - CONFIG_SITE override
+    - configure wrapper (e.g. emconfigure)
+    - make wrapper (e.g. emmake)
+    - additional environment variables
+    - check function to verify SDK
+    """
+
+    name: str
+    pythonexe: str
+    config_site: pathlib.Path
+    configure_wrapper: pathlib.Path
+    make_wrapper: pathlib.Path
+    environ: dict
+    check: Callable[[], None]
+
+    def getenv(self) -> dict:
+        return self.environ.copy()
+
+
+def _check_clean_src():
+    candidates = [
+        SRCDIR / "Programs" / "python.o",
+        SRCDIR / "Python" / "frozen_modules" / "importlib._bootstrap.h",
+    ]
+    for candidate in candidates:
+        if candidate.exists():
+            raise DirtySourceDirectory(candidate, CLEAN_SRCDIR)
+
+
+NATIVE = Platform(
+    "native",
+    pythonexe=sysconfig.get_config_var("BUILDPYTHON"),  # macOS has python.exe
+    config_site=None,
+    configure_wrapper=None,
+    make_wrapper=None,
```

Comment:
Type checkers are unhappy because `pythonexe` can be `None`, while all the other arguments are specified to never be `None`.